src/logplex_api_v3_channels.erl

```diff
@@ -0,0 +1,187 @@
+-module(logplex_api_v3_channels).
+
+-include("logplex.hrl").
+-include("logplex_channel.hrl").
+-include("logplex_logging.hrl").
+
+-export([init/3,
+         rest_init/2,
+         service_available/2,
+         allowed_methods/2,
+         is_authorized/2,
+         resource_exists/2,
+         content_types_provided/2,
+         to_json/2,
+         content_types_accepted/2,
+         from_json/2,
+         delete_resource/2
+        ]).
+
+-record(state, {
+          channel_id :: binary(),
+          tokens = [] :: [logplex_token:token()],
+          drains = [] :: [logplex_drain:drain()]
+         }).
+
+%% @private
+init(_Transport, Req, Opts) ->
+    Route = proplists:get_value(route, Opts),
+    Req1 = logplex_api_v3:prepare(Route, Req),
+    {upgrade, protocol, cowboy_rest, Req1, Opts}.
+
+%% @private
+rest_init(Req, _Opts) ->
+    {ChannelId, Req1} = cowboy_req:binding(channel_id, Req),
+    State = #state{ channel_id = ChannelId },
+    {ok, Req1, State}.
+
+%% @private
+service_available(Req, State) ->
+    logplex_api_v3:service_available(Req, State).
+
+%% @private
+allowed_methods(Req, State) ->
+    {[<<"PUT">>, <<"GET">>, <<"DELETE">>], Req, State}.
+
+%% @private
+is_authorized(Req, State) ->
+    logplex_api_v3:is_authorized(Req, State).
+
+%% @private
+resource_exists(Req, #state{ channel_id = ChannelId } = State) ->
+    %% we need to poll here to wait for nsync replication to catch up
+    Timeout = logplex_app:config(default_redis_poll_ms, 2000),
+    case logplex_channel:poll(ChannelId, Timeout) of
+        Channel when is_record(Channel, channel) ->
+            {ChannelId, Tokens, Drains} = logplex_channel:info(ChannelId),
+            NewState = State#state{ tokens = Tokens,
+                                    drains = Drains },
+            {true, Req, NewState};
+        {error, timeout} ->
+            %% channel was not found
+            {false, Req, State}
+    end.
+
+%% @private
+delete_resource(Req, #state{ channel_id = ChannelId } = State) ->
+    ok = logplex_channel:delete(ChannelId),
+    {true, Req, State}.
+
+%% @private
+content_types_provided(Req, State) ->
+    {[{{<<"application">>, <<"json">>, []}, to_json}], Req, State}.
+
+%% @private
+to_json(Req, #state{ channel_id = ChannelId,
+                     tokens     = Tokens,
+                     drains     = Drains } = State) ->
+    Resp = serialize_channel(ChannelId, Tokens, Drains),
+    {Resp, Req, State}.
+
+%% @private
+content_types_accepted(Req, State) ->
+    {[{{<<"application">>, <<"json">>, []}, from_json}], Req, State}.
+
+%% @private
+from_json(Req, #state{ channel_id = ChannelId,
+                       tokens     = OldTokens,
+                       drains     = Drains } = State) ->
+    {ok, Body, Req1} = cowboy_req:body(Req),
+    try jsx:decode(Body, [return_maps]) of
+        Map ->
+            case validate_payload(Map) of
+                {ok, NewTokenNames} ->
+                    ok = create_or_update_channel(ChannelId),
```

Comment:
This triggers a `logplex_channel:store` to happen, which in turn will overwrite any data stored in redis. Won't this mean that any stored flags like `no_redis` will be obliterated by this operation?